### PR TITLE
Improve GLM-OCR generation and repetition controls

### LINKED ISSUE #1021: GLM-OCR: /chat/completions produces repetition loops on prose; repetition_penalty does not prevent them
## Summary

When OCR'ing a scanned page of flowing English prose through the mlx-vlm server (`/chat/completions`) with `mlx-community/GLM-OCR-bf16`, the model locks into a repetition loop after the first ~500 correct characters and re-emits the same short paragraph dozens of times until `max_tokens` truncates. Setting `repetition_penalty` on the request does not prevent this.

The same source image processed by the same model weights via Ollama (GGUF quantisation, llama.cpp sampler) produces clean, coherent OCR with no loops — so the model itself and image understanding are fine; the issue is isolated to mlx-vlm's generation / sampling path.

## Environment

- mlx-vlm 0.4.4 (installed from `git+https://github.com/Blaizzy/mlx-vlm.git`)
- Python 3.12.13 (Homebrew, macOS arm64)
- macOS Darwin 24.6.0, Apple Silicon Mac Mini
- torch 2.11.0 + torchvision 0.26.0 installed (required for `GlmOcrProcessor` via closed issue #886)
- Model: `mlx-community/GLM-OCR-bf16` (also reproduces with `EZCon/GLM-OCR-mlx`)

## Reproduction

1. Start the server:
   ```
   python -m mlx_vlm.server --trust-remote-code --port 8080
   ```

2. Render a page of flowing English prose at 200 DPI to PNG (image-only scanned PDF, typical size ~300–500 KB). A public-domain Project Gutenberg scan reproduces the issue on any page with an all-caps section heading.

3. POST to `/chat/completions`:
   ```json
   {
     "model": "mlx-community/GLM-OCR-bf16",
     "messages": [{"role": "user", "content": [
       {"type": "image_url", "image_url": {"url": "data:image/png;base64,<...>"}},
       {"type": "text", "text": "OCR this image. Extract ALL text preserving the original formatting, paragraphs, tables, and formulas. Output only the extracted text."}
     ]}],
     "max_tokens": 4096,
     "temperature": 0,
     "repetition_penalty": 1.1
   }
   ```

4. Observe: the first ~500 characters of the response match the page content correctly. Then the model hits a short all-caps heading (e.g. a section header, ~3 words) and begins re-emitting the heading + the subsequent paragraph in a loop until `max_tokens`.

## Measured behaviour

Same 200-DPI rendering, single page of English prose, identical prompt:

| `repetition_penalty` | `temperature` | `top_p` | Output (chars) | Repeated-heading occurrences |
|---|---|---|---|---|
| *unset* | 0.0 | — | 33,688 | ~100 |
| 1.05 | 0.0 | — | 16,870 | 54 |
| 1.1 | 0.0 | — | 16,870 | 54 |
| 1.3 | 0.3 | — | 16,657 | 79 |
| 1.5 | 0.0 | — | 16,594 | 80 |
| 1.1 | 0.5 | 0.9 | 17,086 | 54 |

Ground truth (Ollama GGUF on same image, `glm-ocr:latest`, default sampler): **1,603 chars, heading appears once**. Complete correct transcription of the page.

Observations:
- `repetition_penalty` values from 1.05 to 1.5 do not break the loop on this content
- Increasing `repetition_penalty` from 1.1 → 1.5 at `temperature=0` actually *increased* repeat count (54 → 80)
- Temperature > 0 + `top_p` 0.9 did not help either
- The loop output consistently fills close to `max_tokens` (input ~5000 prompt tokens, output ~4096 generation tokens)

## Hypothesis

Either:
1. The `repetition_penalty` field on `/chat/completions` is accepted by the request model but not threaded into `stream_generate` / `generate_step` for VLM paths, or
2. It is applied but using a codebase convention (e.g. a context window length) that renders it ineffective for this model's tokenisation of repeated all-caps headings, or
3. GLM-OCR's chat template's pre-filled `<think></think>` interacts with sampling in a way that bypasses the penalty.

The counterfactual (Ollama/llama.cpp producing clean output from the same weights on the same image) strongly points at the mlx-vlm sampling path as the isolated cause.

## Related

- #886 (closed) — `GlmOcrProcessor` not loaded by AutoProcessor; same model, different failure mode. Resolved by installing torch+torchvision, but exposed this downstream sampler issue.
- #387 (closed, different model) — similar pattern of mlx-vlm generation diverging from llama.cpp.

## Why this matters

GLM-OCR via mlx-vlm is otherwise an excellent fit for Apple Silicon OCR pipelines (~2× throughput vs. Ollama on benchmarks). This sampler issue makes it unusable today for anyone processing long-form prose; structured content (receipts, forms, equations) appears to mask the problem because short repetitive headers are rarer there.

Happy to test patches against the same reproduction and report back. I'm also skimming the sampling code to see whether a small PR is feasible; will link if I find something.

Thanks for maintaining this project!

### PR DESCRIPTION
**Pull request by Blaizzy**
## Summary
- Build server logits processors from request-level `repetition_penalty`, `presence_penalty`, `frequency_penalty`, and `logit_bias` before inserting requests into `BatchGenerator`.
- Port GLM-OCR image preprocessing and placeholder expansion closer to the Transformers implementation, with processor coverage for the new behavior.
- Fix GLM-OCR M-RoPE position indexing and chunked prefill position reuse so long OCR prompts do not drift into repeated table rows.

## Issue
Fixes #1021

## Validation
- `uv run mlx_vlm.generate --model mlx-community/GLM-OCR-bf16 --prompt "Text Recognition:" --image /Users/prince_canuma/Downloads/table_2.png --temperature 0.0 --no-verbose`
- `uv run python -m compileall -q mlx_vlm/models/glm_ocr/language.py mlx_vlm/tests/test_models.py`
- Focused GLM-OCR RoPE smoke check
- `uv run --with pytest python -m pytest mlx_vlm/tests/test_server.py::TestResponseGenerator::test_server_generation_builds_repetition_logits_processors -q` -> 1 passed
- `uv run --with pytest python -m pytest mlx_vlm/tests/test_server.py -q` -> 94 passed
- `uv run --with pytest --with psutil python -m pytest -q --ignore=mlx_vlm/tests/test_smoke.py` -> 712 passed, 1 skipped

Full bare `pytest -q` still collects `mlx_vlm/tests/test_smoke.py`, whose `test_model_loading` and `test_generation` expect external fixtures (`model_path`, `model`) and error under normal collection.

### CODE REVIEW

**Comment by Blaizzy** [comment]
Validation update for issue #1021:

- Ran `uv run mlx_vlm.generate` against the issue image: `https://github.com/user-attachments/assets/0d98d793-2e29-4ffd-a326-ea1b9605d40b`.
- Confirmed the previous GLM-OCR run was silently ignoring image pixels because the model-specific processor fell back to a tokenizer-only `AutoProcessor` when `torch`/`torchvision` were unavailable.
- Fixed the processor patch so registered model processor failures fail loudly instead of falling back to tokenizer-only behavior.
- Added optional extra `glm-ocr = ["torch", "torchvision"]`; `uv run --extra glm-ocr ...` now loads `GlmOcrProcessor` with `Glm46VImageProcessor`.
- Exposed repetition/presence/frequency penalty controls through the CLI and server schemas.
- Local validation after rebase: `226 passed, 1 skipped` for affected tests; `715 passed, 1 skipped` for the non-smoke suite; `git diff --check` passed.

The GLM-OCR model now receives the image correctly. The long table can still degrade on this input, but using wider-context frequency penalty reduced the hard token-cap repetition in the local run.

**Comment by Blaizzy** [comment]
Correction to my previous validation note:

The `glm-ocr = ["torch", "torchvision"]` extra was the wrong fix. I removed it and replaced the GLM-OCR image path with a local NumPy/PIL `Glm46VImageProcessor` port that follows the Transformers GLM-4V processor logic: GLM `smart_resize`, patch size 14, temporal patch size 2, merge size 2, and the checkpoint image mean/std/size settings from `processor_config.json` / `preprocessor_config.json`.

Validation now runs without torch or torchvision installed:

- `uv pip uninstall torch torchvision`
- `uv run python ...` reports `torch_spec None` and `torchvision_spec None`
- `load("mlx-community/GLM-OCR-bf16")` returns `GlmOcrProcessor` with local `Glm46VImageProcessor`
- The issue image preprocesses to `pixel_values (19824, 1176)`, `image_grid_thw [[1, 168, 118]]`, and 4956 image tokens
- `uv run mlx_vlm.generate --model mlx-community/GLM-OCR-bf16 --image /private/tmp/mlx-vlm-issue-1021/oel_200dpi.png ... --max-tokens 128` starts extracting the table text (`Table III-1c. Group 2B carcinogens`)
- `uv run --with pre-commit pre-commit run --all-files` passed
- affected tests: `228 passed, 1 skipped`
- non-smoke suite: `717 passed, 1 skipped`

Commit: `a80c740`

**Comment by Blaizzy** [comment]
Follow-up: reverted the `mlx_vlm/models/base.py` processor patch behavior change as requested. `base.py` now matches `origin/main` again.

I also removed the test that only existed to enforce that reverted behavior. The GLM-OCR local NumPy image processor remains in place, so GLM-OCR no longer depends on adding `torch`/`torchvision`.

Validation before push:
- `uv run --with pre-commit pre-commit run --all-files` passed
- affected suite: `227 passed, 1 skipped`
- non-smoke suite: `716 passed, 1 skipped`

Commit: `eab44b7`

**Comment by Blaizzy** [comment]
Follow-up: made GLM-OCR preprocessing bit-exact with the upstream Transformers GLM46V image processor when the Transformers torch/torchvision backend is available.

Implementation keeps the NumPy/PIL path as a fallback, so torch/torchvision are still not package dependencies. When torch/torchvision are present, `Glm46VImageProcessor` delegates preprocessing to the same `transformers.models.glm46v.image_processing_glm46v.Glm46VImageProcessor` backend used by `zai-org/GLM-OCR`, then converts tensors back to NumPy for mlx-vlm.

Numerical check on `/Users/prince_canuma/Downloads/table_2.png`:
- `pixel_equal True`
- `grid_equal True`
- `max_abs 0.0`

Validation:
- `uv run --with pre-commit pre-commit run --all-files` passed
- affected suite: `227 passed, 2 skipped`
- non-smoke suite: `716 passed, 2 skipped`

Commit: `698faf4`

**Comment by Blaizzy** [comment]
Follow-up for the GLM-OCR repetition loop:

The exact command now stops cleanly before emitting the duplicate table row that starts the loop:

`uv run mlx_vlm.generate --model mlx-community/GLM-OCR-bf16 --prompt "Text Recognition:" --image /Users/prince_canuma/Downloads/table_2.png --temperature 0.0`

Before: generated to the 2048-token cap and repeated the Benzofuran/Benzophenone/Benzoyl chloride block.
After: stops at 489 generated tokens, ending after the first `Auramine (technical grade)` row, with no repeated loop or dangling duplicate prefix.

Implementation: for GLM-OCR greedy decoding only, buffer OCR output by line and stop before emitting a duplicate completed line; also keep the token repeated-tail guard as a fallback. This does not change logits/sampling and does not affect other model types.

Validation:
- exact command run locally: no loop, 489 generated tokens
- `uv run --with pre-commit pre-commit run --all-files` passed
- affected suite: `231 passed, 2 skipped`
- non-smoke suite: `720 passed, 2 skipped`

Commit: `c1b09c7`

**Comment by Blaizzy** [comment]
Reverted the GLM-OCR generation-side repeated row loop guard in d72b766 at maintainer request. The bit-exact GLM-OCR processor/preprocessing changes remain in the branch.
